router/core/supervisor.go

```diff
@@ -0,0 +1,186 @@
+package core
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/wundergraph/cosmo/router/pkg/config"
+	"go.uber.org/zap"
+)
+
+// RouterSupervisor is a supervisor for the router.
+type RouterSupervisor struct {
+	logger *zap.Logger
+
+	router       *Router
+	routerCtx    context.Context
+	routerCancel context.CancelFunc
+
+	configPath string
+
+	// Sending to this channel will trigger a graceful shutdown of the router
+	// Sending true will result in a Reload, false will result in a Shutdown
+	// Value means "to kill or not to kill"
+	shutdownChan chan bool
+
+	lifecycleHooks *LifecycleHooks
+
+	resources *RouterResources
+}
+
+// LifecycleHooks is a struct for holding router lifecycle hooks.
+type LifecycleHooks struct {
+	PreCreate func(*RouterResources) error
+}
+
+// RouterResources is a struct for holding resources used by the router.
+type RouterResources struct {
+	Config *config.Config
+	Logger *zap.Logger
+}
+
+// RouterSupervisorOpts is a struct for configuring the router supervisor.
+type RouterSupervisorOpts struct {
+	ConfigPath     string
+	BaseLogger     *zap.Logger
+	LifecycleHooks *LifecycleHooks
+}
+
+// NewRouterSupervisor creates a new RouterSupervisor instance.
+func NewRouterSupervisor(opts *RouterSupervisorOpts) *RouterSupervisor {
+	return &RouterSupervisor{
+		shutdownChan:   make(chan bool),
+		logger:         opts.BaseLogger.With(zap.String("component", "supervisor")),
+		lifecycleHooks: opts.LifecycleHooks,
+		configPath:     opts.ConfigPath,
+		resources: &RouterResources{
+			Logger: opts.BaseLogger,
+		},
+	}
+}
+
+func (rs *RouterSupervisor) createRouter() error {
+	// Provide a way to cancel all running components of the router after graceful shutdown
+	// Don't use the parent context that is canceled by the signal handler
+	routerCtx, routerCancel := context.WithCancel(context.Background())
+
+	// TODO: Test if this actually allows router failure and recovery
+	router, err := newRouter(routerCtx, RouterCreateParams{
+		Config: rs.resources.Config,
+		Logger: rs.resources.Logger,
+	})
+	if err != nil {
+		routerCancel()
+		return fmt.Errorf("failed to create router: %w", err)
+	}
+
+	rs.router = router
+	rs.routerCtx = routerCtx
+	rs.routerCancel = routerCancel
+
+	return nil
+}
+
+func (rs *RouterSupervisor) startRouter() error {
+	if err := rs.router.Start(rs.routerCtx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (rs *RouterSupervisor) stopRouter() error {
+	// Enforce a maximum shutdown delay to avoid waiting forever
+	// Don't use the parent context that is canceled by the signal handler
+	shutdownCtx, cancel := context.WithTimeout(context.Background(), rs.resources.Config.ShutdownDelay)
+	defer cancel()
+
+	rs.logger.Info("Graceful shutdown of router initiated", zap.String("shutdown_delay", rs.resources.Config.ShutdownDelay.String()))
+
+	if err := rs.router.Shutdown(shutdownCtx); err != nil {
+		return fmt.Errorf("failed to shutdown router gracefully: %w", err)
```

Comment:
Please verify with the old implementation. Logs are missing.

Reply:
```
18:16:03 PM INFO core/supervisor.go:100 Graceful shutdown of router initiated {"hostname": "june.jlt.sh", "pid": 71083, "service": "@wundergraph/router", "service_version": "dev", "component": "supervisor", "shutdown_delay": "1m0s"}
```

What do you mean missing?

Reply:
The location will have changed and it will have `component: supervisor`, is that an issue?

Reply:
This

```go
	if err = router.Shutdown(shutdownCtx); err != nil {
		if errors.Is(err, context.DeadlineExceeded) {
			logger.Warn("Router shutdown deadline exceeded. Consider increasing the shutdown delay")
		}
		logger.Fatal("Could not shutdown router gracefully", zap.Error(err))
	} else {
		logger.Info("Router shutdown successfully")
	}

	logger.Debug("Router exiting")
```